File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> last(256,-1);
        int start = 0;
        int maxlen = 0;
        for(int i = 0;i < s.size();i++){
            char ch = s[i];
            if(last[ch] >= start) start = last[ch] + 1;
            last[ch] = i;
            maxlen = max(maxlen,i - start + 1);
        }
        return maxlen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
     int start = 0;
     int end = 0;
     int maxlen = Math.min(s.length(),1);
     Set<Character> seen = new HashSet<>();
     while(end < s.length()){
        char c = s.charAt(end);
        while(seen.contains(c)){
            seen.remove(s.charAt(start));
            start++;
        }
        seen.add(c);
        int size = end - start + 1;
        maxlen = Math.max(maxlen,size);
        end++;
    
     }
     return maxlen;   
    }

}